docworks-dts/lib/multiple-files/typeResolver.js

```diff
@@ -0,0 +1,159 @@
+const camelCase_ = require('lodash/camelCase')
+const { addDependency } = require('./providers/modulesDependencies')
+const { $W_NAME, TYPES } = require('./constants')
+const {
+	BASE_TYPES,
+	dtsUnion,
+	dtsIntersection,
+	dtsNamedTypeReference
+} = require('./generators')
+
+const isString = value => typeof value === 'string'
+const hasNameProperty = type => typeof type === 'object' && type.name
+const isListOfTypes = type => Array.isArray(type) && type.length > 0
+const isInvalidOptions = ({ intersection, union }) =>
+	(intersection && union) || (!intersection && !union)
+
+// Array of more one value defines for us a union type.
+const isUnionType = docsTypes => {
+	return docsTypes.length > 1
+}
+
+// Supported key value types: Record/Map.
+const isKeyValueType = typeName => {
+	return [TYPES.RECORD, TYPES.MAP].some(t => t === typeName)
+}
+
+// Docs Type Definition: https://github.com/wix-private/p13n/blob/master/wixplorer/wix-docs-server/proto/docs/README.md
+const isNewDocsTypesDefinition = type =>
+	type.some(
+		({ nativeType, complexType, referenceType }) =>
+			nativeType || complexType || referenceType
+	)
+
+const isRootMember = (type, { name }) => type.startsWith(`${name}.`)
+
+const isInnerMember = (type, { name, memberOf }) =>
+	memberOf && type.includes(`${name}.`) && type.includes(`${memberOf}.`)
+
+const isSameModule = (type, { memberOf }) => {
+	if (!memberOf) return false
+	const [memberOfRootModule] = memberOf.split('.')
+	const [typePartsRootModule] = type.split('.')
+	return type.includes('.') && memberOfRootModule === typePartsRootModule
+}
+
+const isCrossReferenceType = (type, service) =>
+	type.includes('.') && !isSameModule(type, service)
+
+const resolveCrossReferenceType = (type, service) => {
+	// do nothing if the referenced type is a member of $w because $w is global
+	if (type.startsWith($W_NAME)) return type
+
+	const { name, memberOf } = service
+	const [currentModuleName] = memberOf ? memberOf.split('.') : [name]
+	const [crossedTypeModuleName, ...otherParts] = type.split('.')
+
+	addDependency(currentModuleName, crossedTypeModuleName)
+	return `${camelCase_(crossedTypeModuleName)}.${otherParts.join('.')}`
+}
+
+const fixType = (type, service) => {
+	if (isString(type)) {
+		const { name: parentServiceName } = service
+		if (isRootMember(type, service)) {
+			return type.replace(`${parentServiceName}.`, '')
+		}
+		if (isInnerMember(type, service)) {
+			const splitedParts = type.split(`${service.memberOf}.`)
+			return `${splitedParts[splitedParts.length - 1]}`
+		}
+		if (isSameModule(type, service)) {
+			const [rootModule] = type.split('.')
```

Comment:
Can't this just be `service.memberOf`?

Reply:
no, I want to fetch only the root module and in some cases, memberOf can point to an inner module like: `ROOT_MODULE.SUB.SUB `